### PR TITLE
Add a nlls trait to BVProblem

### PR DESCRIPTION
**Pull request by avik-pal**
## Checklist

- [ ] Appropriate tests were added
- [x] Any code changes were done in a way that does not break public API
- [x] All documentation related to code changes were updated
- [x] The new code follows the
  [contributor guidelines](https://github.com/SciML/.github/blob/master/CONTRIBUTING.md), in particular the [SciML Style Guide](https://github.com/SciML/SciMLStyle) and
  [COLPRAC](https://github.com/SciML/COLPRAC).
- [x] Any new documentation only uses public API
  
## Additional context

The main compilation cost for BVDiffEq is coming from solving NLLS problems. Having this allows us to have a type stable path for NLLS BVPs.

### CODE REVIEW

File: src/problems/bvp_problems.jl [hunk 1]
```diff
@@ -115,18 +124,50 @@ struct BVProblem{uType, tType, isinplace, P, F, PT, K} <:
...
+                    end
+                end
+            else
+                if f.bcresid_prototype !== nothing
+                    _nlls = length(f.bcresid_prototype) != length(_u0)
+                else
+                    if iip
+                        _nlls = false # Should we assume `true` instead?
+                    else
+                        _nlls = length(f.bc(FFakeSolutionObject(u0), p, tspan)) != length(_u0)
```
Comment (ErikQQY) [inline src/problems/bvp_problems.jl:162]: Should be ```FakeSolutionObject```
```suggestion
                        _nlls = length(f.bc(FakeSolutionObject(u0), p, tspan)) != length(_u0)
```

File: src/solutions/ode_solutions.jl [hunk 2]
```diff
@@ -115,6 +115,8 @@ struct ODESolution{T, N, uType, uType2, DType, tType, rateType, P, A, IType, S,
     stats::S
     alg_choice::AC
     retcode::ReturnCode.T
+    resid::R
+    original::O
```
Comment (ErikQQY) [inline src/solutions/ode_solutions.jl:119]: What about creating a ```BVPSolution``` to specifically handle these details only in the BVP context?

Comment (avik-pal) [reply src/solutions/ode_solutions.jl:119]: I am open to that. We can do that by subtyping AbstractODESolution.

File: src/solutions/ode_solutions.jl [hunk 3]
```diff
@@ -117,6 +117,88 @@ struct ODESolution{T, N, uType, uType2, DType, tType, rateType, P, A, IType, S,
...
+    interp::IType
+    dense::Bool
+    tslocation::Int
+    stats::S
+    alg_choice::AC
+    retcode::ReturnCode.T
+    resid::R
+    original_ivp::OI
+    original_nlsolve::ON
+end
```
Comment (avik-pal) [inline src/solutions/ode_solutions.jl:179]: @ErikQQY thoughts on this?

Comment (ErikQQY) [reply src/solutions/ode_solutions.jl:179]: 👍I think with this we can specifically handle solutions in the BVP context

File: src/solutions/ode_solutions.jl (hunk 2 above)
Comment (ChrisRackauckas) [reply src/solutions/ode_solutions.jl]: IMO it just seems a bit wasteful code wise. It's just two more fields, and `original` makes sense in the IVP case as well (we should use it with ODEInterface)

**Comment by avik-pal** [comment]
I am happy with the current state of this. To summarize

1. Adds an `nlls` trait to BVProblem -- we try to infer it but err on the side of caution and set it to Nothing if we can reliably figure it out. It is the job of the upstream library to figure out how to deal with this case
2. Adds `original` to the ODESolution. I agree with @ChrisRackauckas that we probably don't want to unnecessarily copy the code into a BVPSolution since original might be useful for ODEs as well. `resid` not for IVPs but that is probably not a good enough reason to copy so much code.

**Review by ChrisRackauckas — APPROVED** [review]

**Pull request merged**